### ISSUE TITLE
Preloading assets in Rollup apps

### ISSUE THREAD

**Issue by Rich-Harris** [sveltejs/sapper#408]
Currently, only `client.xyz123.js` is preloaded. Sapper should add preload headers for all the JS and CSS needed for the page being server-rendered

**Comment by MaxMilton** [comment]
I'd like to suggest we also inject `<link href=xxx rel=preload as=xxx>` and `<link href=xxx rel=prefetch as=xxx>` into the document head for `sapper export` so static builds can also benefit. The headers are great since some systems can use them for HTTP/2 push but the reality is it's still early days for push.

**Comment by rwwagner90** [comment]
Any chance we could get one of the PRs merged in with the fix? I would love to increase my Lighthouse score 😃

**Comment by ahfarmer** [comment]
Static sites can benefit from link headers as well, depending on how they are deployed. For instance with Netlify you can create a _headers file which specifies the headers for each path: https://www.netlify.com/docs/headers-and-basic-auth/

Would be great if we could export the headers somehow, or if there is a place to hook in a plug-in to create the _headers file.

**Comment by benmccann** [comment]
I don't think adding a `link` `prefetch` would help, especially once https://github.com/sveltejs/sapper/pull/1123 is merged, because those assets will be discovered early in the page anyway. The headers are valuable though because the preload spec mentions that servers may initiate a PUSH when they see a preload header and most CDNs have standardized around this behavior to initiate a PUSH.

**Comment by benmccann** [comment]
`build.ts` uses the `build_info` to create the headers. `RollupResult.to_json` only returns `client.js`. We should probably make it return the whole dependency tree.

The key thing here seems to be fixing this TODO:
https://github.com/sveltejs/sapper/blob/857a3e786ef3bab13445c1cae685c7493d3b7062/src/core/create_compilers/RollupResult.ts#L39

`compiler.chunks[i].modules.dynamicImports` should be preload based on the page URL. original file path available in `facadeModuleId`
`compiler.chunks[i].modules.imports` should be preloaded the if `filename` is in the `imports` dependency tree of page that was dynamically imported. `filename` should be unique since it contains a hash

**Comment by benmccann** [comment]
`create_app.ts` builds the client manifest and then _I think_ Rollup reprocesses it

So what would start out as:
```
		js: () => import("../../../routes/index.svelte"),
```

Would get reprocessed to:
```
		js: () => import('./index.efebda61.js'),
```

And I think Rollup only finds and preprocesses the path if it's used in an `import` statement

My initial thought was that we could have `create_app` (or something similar) generate a new file that provides the mapping from route to svelte component. Because we really want the path and not an import statement, but that means we get the before path (i.e. the `.svelte` and not the `.js`). We could them combine that with the `compiler.chunks` above. I'm not sure if a generated file would be available at the time we're utilizing `RollupResult`? Though JavaScript has dynamic imports, so perhaps if we imported it a later on in the compilation process and not at the top of the file it would work. But maybe this is adding too much logic into `RollupCompiler` at that point

I think the more of this info we can get into `manifest-server.mjs` the better

**Comment by benmccann** [comment]
I made a small library to extract the dependency tree from the Rollup `OutputChunk[]`: https://github.com/benmccann/rollup-dependency-tree

**Comment by benmccann** [comment]
I sent a PR that implements this for JS files when using Rollup: https://github.com/sveltejs/sapper/pull/1269

We could probably implement it for CSS files and Webpack still, but it's a pretty good start and probably the most meaningful piece to get done.

**Comment by benmccann** [comment]
@Rich-Harris is Sapper supposed to do code splitting on CSS files? It seems like rollup-plugin-svelte may be broken with Rollup 2 https://github.com/sveltejs/sapper-template/issues/233. I can't make further progress on this without some advice from you or maybe @Conduitry or @antony on how the CSS is intended to work and probably someone with more Rollup experience fixing that issue assuming it is indeed an issue

**Comment by antony** [comment]
@benmccann interesting, I'd not noticed a difference. Ill check as soon as I'm able.

**Comment by kevmodrome** [comment]
Not sure if relevant but might this be something to take a look at: https://github.com/vikerman/rollup-plugin-hoist-import-deps ?

**Comment by benmccann** [comment]
@kevmodrome thanks for the pointer. I'd actually taken a look at that plugin. It's helpful, but https://github.com/sveltejs/sapper/pull/1269 does even better. You can see my findings here: https://github.com/sveltejs/sapper/issues/1260#issuecomment-642289567.

Anyways, that plugin and the work I've done so far just address the JS piece. My remaining question is about the CSS. I'm happy to put together a PR address the CSS piece as well, but think we should figure out the Rollup 2 compatibility first

**Comment by habibrosyad** [comment]
Just as an addition to issue https://github.com/sveltejs/sapper-template/issues/233 pointed by @benmccann, I had actually posted another issue in https://github.com/rollup/rollup/issues/3655 along with the repro details.

**Comment by benmccann** [comment]
With much help from the Rollup team, I was able to fix CSS code splitting in Sapper for Rollup 2 and sent a PR for it: https://github.com/sveltejs/sapper/pull/1306. For Rollup 2 users will need to use Rollup 2.21.0+ and will need https://github.com/sveltejs/sapper/pull/1306 merged in order to get code splitting. Otherwise everything is loaded as one big CSS file

With that fixed, I've gone ahead and updated my PR https://github.com/sveltejs/sapper/pull/1269 to `preload` CSS which was fairly straight-forward since most of the work was already done and we just needed to connect some pieces. It will work with any version of Rollup 1 and Rollup 2, but you will need the fix I just mentioned and latest version of Rollup to get code splitting if using Rollup 2

**Comment by LucianVoju** [comment]
Hello, got any update when this merge will happen? Thx

**Comment by LucianVoju** [comment]
Hello, are there any needed configurations in Rollup or Sapper to benefit from this?

**Comment by benmccann** [comment]
No configuration is necessary

**Comment by LucianVoju** [comment]
Any package update?

**Comment by benmccann** [comment]
Yes, you need to use Sapper 0.28.0

**Comment by LucianVoju** [comment]
ok, thank you

**Comment by LucianVoju** [comment]
got this message:
'preload' is not exported by 'src/routes/_layout.svelte'
333:
334:   if (!root_preloaded) {
335:     const root_preload = root_comp.preload || (() => {});
                                        ^
336:     root_preloaded = initial_data.preloaded[0] || root_preload.call(preload_context, {
337:       host: page.host,

**Comment by benmccann** [comment]
That's unrelated to this PR. Please see the migration guide: https://sapper.svelte.dev/migrating#0_27_to_0_28

**Comment by LucianVoju** [comment]
Did all the  steps and still have the message, also on the [realworld](https://github.com/sveltejs/realworld) project after sapper update getting the same message

**Comment by benmccann** [comment]
It's the last bullet in the migration guide. It works fine on the realworld project. See my PR that's pending there to do the upgrade

**Comment by LucianVoju** [comment]
All is fine, thank you.
